sentinel-core/src/test/java/com/alibaba/csp/sentinel/log/ConsoleHandlerTest.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright 1999-2018 Alibaba Group Holding Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.alibaba.csp.sentinel.log;
+
+import org.junit.Test;
+
+import java.util.logging.Level;
+import java.util.logging.LogRecord;
+
+import static org.junit.Assert.*;
+import static org.mockito.Mockito.*;
+
+/**
+ * Test cases for {@link ConsoleHandler}.
+ *
+ * @author cdfive
+ */
+public class ConsoleHandlerTest {
+
+    @Test
+    public void testPublish() {
+        ConsoleHandler consoleHandler = new ConsoleHandler();
+
+        LogRecord logRecord = mock(LogRecord.class);
+        when(logRecord.getMillis()).thenReturn(System.currentTimeMillis());
+
+        when(logRecord.getLevel()).thenReturn(Level.INFO);
+        when(logRecord.getMessage()).thenReturn("test info message");
+        consoleHandler.publish(logRecord);
+        assertEquals(consoleHandler.getStdoutHandler(), consoleHandler.getCurrentHandler());
+
+        when(logRecord.getLevel()).thenReturn(Level.WARNING);
+        when(logRecord.getMessage()).thenReturn("test warning message");
+        consoleHandler.publish(logRecord);
+        assertEquals(consoleHandler.getStderrHandler(), consoleHandler.getCurrentHandler());
```

Comment:
If you do unit testing you can simply mock the `System.out` & `System.err` for correct response.

And I think `LogRecord` has now need to be mocked because we can easily get an instance we need.

Reply:
> LogRecord has now need to be mocked because we can easily get an instance we need.
Good idea.

How to mock **System.out**? I found a way by using `ByteArrayOutputStream` like this:
```java
ByteArrayOutputStream baosOut = new ByteArrayOutputStream();
System.setOut(new PrintStream(baosOut));
```
And then verify `baosOut.toString()`, is it ok?

PS: Thanks for so detailed reviewing!

Reply:
Kinds of, for example:

```java
    @Test
    public void test() throws IOException {
        PrintStream stdout = mock(PrintStream.class);
        PrintStream stderr = mock(PrintStream.class);
        System.setOut(stdout);
        System.setErr(stderr);
        System.err.write("asdf\n".getBytes());
        System.out.write("asdf\n".getBytes());
        System.out.write("asdf\n".getBytes());
        verify(stdout, times(2)).write(any(byte[].class));
        verify(stderr).write(any(byte[].class));
    }
```

Reply:
Thanks, I learned it. It can verify action.

Reply:
As encapsulated by `OutputStreamWriter`, to verify the byte[] write is not very convenient.
I tried with `ByteArrayOutputStream`, it works and no mock in this test case.Is it OK?

Reply:
> As encapsulated by `OutputStreamWriter`, to verify the byte[] write is not very convenient.
> I tried with `ByteArrayOutputStream`, it works and no mock in this test case.Is it OK?

Yes there is no need to verify the content i think we only try to verify the times invoked and it's enough.